haystack/squad_data.py

```diff
@@ -0,0 +1,279 @@
+        df_aggregated_questions = pd.merge(df_aggregated_questions, questions)
+
+        logger.info("Aggregating the paragraphs of each document")
+        df_grouped_paragraphs = df_aggregated_questions.groupby(["title"])
+        df_aggregated_paragraphs = df[["title"]].drop_duplicates().reset_index()
+        paragraphs = df_grouped_paragraphs.progress_apply(self._aggregate_passages).rename("paragraphs")
+        paragraphs = pd.DataFrame(paragraphs).reset_index()
+        df_aggregated_paragraphs = pd.merge(df_aggregated_paragraphs, paragraphs)
+
+        df_aggregated_paragraphs = df_aggregated_paragraphs[["title", "paragraphs"]]
+        ret = df_aggregated_paragraphs.to_dict("records")
+
+        return ret
+
+    @staticmethod
+    def _aggregate_passages(x):
+        x = x[["context", "qas"]]
+        ret = x.to_dict("records")
+        return ret
+
+    @staticmethod
+    def _aggregate_questions(x):
+        x = x[["question", "id", "answers", "is_impossible"]]
+        ret = x.to_dict("records")
+        return ret
+
+    @staticmethod
+    def _aggregate_answers(x):
+        x = x[["answer_text", "answer_start"]]
+        x = x.rename(columns={"answer_text": "text"})
+
+        # Span anwser
+        try:
+            x["answer_start"] = x["answer_start"].astype(int)
+            ret = x.to_dict("records")
+
+        # No answer
+        except ValueError:
+            ret = []
+
+        return ret
+
+    def set_data(self, data):
+        self.data = data
+        self.df = self.to_df(data)
+
+    def sample_questions(self, n):
+        """
+        Return a sample of n questions in SQuAD format (list of SQuAD document dictionaries)
+        Note, that if the same question is asked on multiple different passages, this fn treats that
+        as a single question
+        """
+        all_questions = self.get_all_questions()
+        sampled_questions = random.sample(all_questions, n)
+        df_sampled = self.df[self.df["question"].isin(sampled_questions)]
+        return self.df_to_data(df_sampled)
+
+    def get_all_paragraphs(self):
+        """Return all paragraph strings"""
+        return self.df["context"].unique().tolist()
+
+    def get_all_questions(self):
+        """Return all question strings. Note that if the same question appears for different paragraphs, it will be
+        returned multiple times by this fn"""
+        df_questions = self.df[["title", "context", "question"]]
+        df_questions = df_questions.drop_duplicates()
+        questions = df_questions["question"].tolist()
+        return questions
+
+    def get_all_document_titles(self):
+        """Return all document title strings"""
+        return self.df["title"].unique().tolist()
+
+if __name__ == "__main__":
+    filename1 = "../data/squad20/train-v2.0.json"
+    filename2 = "../data/squad20/dev-v2.0.json"
+
+    # Load file1 and take a sample of 10000 questions
+    sd = SquadData.from_file(filename1)
```

Comment:
Could we use [FARM's functionality](https://github.com/deepset-ai/FARM/blob/816b4e3e65c142f8a31a63833058b75fe0419ed4/farm/data_handler/utils.py#L300) to download squad data if the files do not exist yet or would that be too complicated? I understand this is only a small example.

Reply:
Good idea!